### PR TITLE
Frontend batching

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Batches jobs in the requests from the query-frontend queue to the queriers. Previously, the frontend would send each job 1 at at time with an individual http request. This PR adds a configurable parameter to allow the frontend to send more than one request at once. 

**Other changes:**
- Docs of course! Including an update to the search performance tuning doc with some more current information.
- Adds a new histogram metric `tempo_query_frontend_actual_batch_size` to track the actual size of the batches being farmed to the queriers
- Better testing of the queues and frontend worker.
- Added the ability for the querier to signal to the frontend the features it supports for seamless rollouts.

**Performance testing**
The goal with the setup was to create a cluster that could execute the 36k jobs created by the test query simultaneously. This way job throughput from frontend -> querier could be tested more directly.

  - 80 queriers
  - 500 jobs per querier
  - Total cluster capacity 40k jobs
  - No reliance on serverless

**Results**
```
batch size    overall query latency     p99 job time in queue
1             8.5s                      4.9s
2             7.6s                      2.4s
5             6.7s                      1s
10            9s                        4.4s
1*            9.6s                      9s

*current image
```

The overall latency of queries where total jobs > total cluster capacity was not as impressively reduced, but this is a good step in the right direction.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by mdisibio — APPROVED** [review]
This looks great and I like the way it is controlled by querier features. A few small q's, but none blocking and will go ahead and approve.

File: modules/frontend/v1/request_batch.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	if len(b.pipelineRequests) == 1 {
+		return b.pipelineRequests[0].originalCtx.Done()
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for _, r := range b.pipelineRequests {
+			select {
+			case <-r.originalCtx.Done():
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/frontend/v1/request_batch.go:64]: will add a comment :+1:

File: modules/querier/worker/frontend_processor.go [hunk 2]
```diff
@@ -93,25 +96,58 @@ func (fp *frontendProcessor) process(c frontendv1pb.Frontend_ProcessClient) erro
...
-func (fp *frontendProcessor) runRequest(ctx context.Context, request *httpgrpc.HTTPRequest, sendHTTPResponse func(response *httpgrpc.HTTPResponse) error) {
+func (fp *frontendProcessor) runRequests(ctx context.Context, requests []*httpgrpc.HTTPRequest) []*httpgrpc.HTTPResponse {
+	wg := sync.WaitGroup{}
+
+	responses := make([]*httpgrpc.HTTPResponse, len(requests))
+	for i, request := range requests {
+		wg.Add(1)
+		go func(i int, request *httpgrpc.HTTPRequest) {
+			responses[i] = fp.runRequest(ctx, request)
+			wg.Done()
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/querier/worker/frontend_processor.go:143]: nope. it was probably just so simple i wrote it that way. since defer is a more standard/g/foolproof way to go about this i will defer

[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/querier/worker/frontend_processor.go:143]: you could say that i'm deferring to defer

File: modules/frontend/v1/frontend.go [hunk 3]
```diff
@@ -203,43 +214,62 @@ func (f *Frontend) Process(server frontendv1pb.Frontend_ProcessServer) error {
 
 	lastUserIndex := queue.FirstUser()
 
+	reqBatch := &requestBatch{}
+	batchSize := 1
+	if querierSupportsBatching(querierFeatures) {
+		batchSize = f.cfg.MaxBatchSize
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/frontend/v1/frontend.go:220]: Initially I wrote it so that the querier told the frontend a batch size. There are two places this can occur. 

1) In the initial TYPE_GET_ID request
2) In the response to every TYPE_HTTP_REQUEST*

I eventually backed off of it b/c even in the case of #2 the batch size will lag the actual state of the querier. Every GRPC streaming connection to the querier would have a possibly different batch size that would almost always be based on out of date information.

I like the ideal, but I'm concerned about the complexity and failure modes. I also considered an out of band streaming call that regularly updated the batch size between the querier <-> frontend globally for all streaming connections, but that seemed outside the scope of this PR.

In short, I think it's a good idea, but may take a bit to get there. I see this is a good first step in a more complete solution like that.

File: modules/frontend/v1/frontend.go [hunk 4]
```diff
@@ -96,6 +98,11 @@ func New(cfg Config, limits Limits, log log.Logger, registerer prometheus.Regist
 			Help:    "Time spend by requests queued.",
 			Buckets: prometheus.DefBuckets,
 		}),
+		actualBatchSize: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+			Name:    "tempo_query_frontend_actual_batch_size",
+			Help:    "Batch size.",
+			Buckets: prometheus.LinearBuckets(1, 1, 10),
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/frontend/v1/frontend.go:104]: yup! i meant to do that but obviously it slipped my mind. Good call!

[the remark this answers is not in the document]
Comment (mdisibio) [reply modules/frontend/v1/frontend.go:104]: 👍  The config check below needs to be relocated up then.

File: modules/frontend/v1/frontend.go [hunk 5]
```diff
@@ -254,24 +284,48 @@ func (f *Frontend) Process(server frontendv1pb.Frontend_ProcessServer) error {
...
-		// downstream req.  Only way we can do that is to close the stream.
-		// The worker client is expecting this semantics.
-		case <-req.originalCtx.Done():
-			return req.originalCtx.Err()
-
-		// Is there was an error handling this request due to network IO,
-		// then error out this upstream request _and_ stream.
-		case err := <-errs:
-			req.err <- err
+		err = reportResponseUpstream(reqBatch, errs, resps)
```
Comment (zalegrala) [inline modules/frontend/v1/frontend.go:287]: Do we have a context to pass?  Wondering if it might simplify the context handling below.

**Review by zalegrala — APPROVED** [review]
This looks pretty good to me, nice improvement.  Will be interesting to see the results on the dashboard.  Had a question about the context handling but not blocking.

File: modules/frontend/v1/frontend.go (hunk 5 above)
Comment (joe-elliott) [reply modules/frontend/v1/frontend.go:289]: If the streaming GRPC server connection itself drops or context is cancelled then the `.Send()` returns an error and this case is hit:
https://github.com/grafana/tempo/pull/2677/files#diff-0914703aed52090bd72851004df203444207d9d48677c10860b0459afef1a0b9R311

If the request is cancelled upstream then this case is hit:
https://github.com/grafana/tempo/pull/2677/files#diff-0914703aed52090bd72851004df203444207d9d48677c10860b0459afef1a0b9R304

If the requests are cancelled downstream then we get an http response and this case is hit:
https://github.com/grafana/tempo/pull/2677/files#diff-0914703aed52090bd72851004df203444207d9d48677c10860b0459afef1a0b9R304

I think everything is covered.
